### PR TITLE
update(ci): add update emoji schedule workflow

### LINKED ISSUE #2434: Workflow: Setup a daily sync scheduler to sync emoji data instead of run in every build
## Enhancement
Details in the discussions in #2291 that the `build:emoji` may not need run in every build.
As a workaround solutions, we could setup a daily checker to sync the emojiData changes for now.
Besides, when we finish the #2431, we could move this workflow to the plugin part also, if needs to.

### PR DESCRIPTION
**Pull request by Koooooo-7**
## Summary

Add update emoji schedule workflow to daily update emoji data.
Changes:
- Support schedule at `UTC 2:00` daily run,  or manually trigger.
- Remove the `build:emoji` out of `build`.
- Remove the `try catch` from `emoji.js` to ensure the `writeJS` and `wirteMarkdown` in same `transaction` （both success or failed), so, we could just check the last write file (`src/core/render/emoji-data.js`) changes and add them all.
- The commit branch uses the `update-emoji-data-$current_date-${last commit hash 7}` to be an unique branch commit, it makes it possible that we could manually trigger multi times in a day or it were be trigger both in schedule and manually does.

Others:
- Hasn't added a `build:emoji` in `release.sh` since it is unlikely happen as per our discussion.

close #2434 

- PR 
Screenshot 2024-05-22 at 22 42 06

- Branch details
Screenshot 2024-05-22 at 22 42 47

## What kind of change does this PR introduce?

  Build-related changes
## For any code change,

- [ ] Related documentation has been updated, if needed
- [x] Related tests have been added or updated, if needed

## Does this PR introduce a breaking change?

No

## Tested in the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: .github/workflows/rotate-emoji.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
+name: Rotate emojiData
```
Comment (jhildenbiddle) [inline .github/workflows/rotate-emoji.yml:1]: Minor nitpick, but "Update Emoji" (YAML `name`) and `update-emoji.yml` (file name) are probably better names since we aren't "rotating" emoji in/out.

File: .github/workflows/rotate-emoji.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+on:
+  schedule:
+    - cron: '0 2 * * *'
+  workflow_dispatch:
+
+concurrency:
+  group: rotate-emoji-group
```
Comment (jhildenbiddle) [inline .github/workflows/rotate-emoji.yml:9]: `rotate-emoji-group` => `update-emoji-group`

File: .github/workflows/rotate-emoji.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+jobs:
+  sync-emojis-from-github:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Setup Node.js v20
+        uses: actions/setup-node@v3
```
Comment (jhildenbiddle) [inline .github/workflows/rotate-emoji.yml:18]: Why not [`actions/setup-node@4`](https://github.com/actions/setup-node/releases)?

**Review by jhildenbiddle — CHANGES_REQUESTED** [review]
This looks great, @Koooooo-7. A few minor naming nitpicks, but otherwise this looks good.

Has this been tested? If not, it seems like we could by removing a `data` property entry from `/src/core/render/emoji-data.js`, committing the change, then waiting to see if this workflow generates the expected PR.

File: .github/workflows/rotate-emoji.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        run: |
+          if git diff --quiet -- "src/core/render/emoji-data.js"; then
+            echo "No changes Found. Skipping PR creation."
+          else
+            echo "Changes detected in emoji-data.js. Creating PR."
+
+            current_date=$(date +'%Y-%m-%d')
+
+            branch_name="update-emoji-data-$current_date"
+            pr_title="update(emoji): sync emojiData and find update emojis in [$current_date]"
```
Comment (jhildenbiddle) [inline .github/workflows/rotate-emoji.yml:39]: The grammar is a little hard to follow in some of the titles. How about the following:

```yaml
pr_title="update(emoji): Sync emoji data with GitHub emoji API [$current_date]"
```
```yaml
git commit -m "update(emoji): Sync emoji data with GitHub emoji API [$current_date]"
```

File: .github/workflows/rotate-emoji.yml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+      - name: Fetch emojiData
+        run: npm run build:emoji
+
+      - name: Check for changes
+        run: |
+          if git diff --quiet -- "src/core/render/emoji-data.js"; then
+            echo "No changes Found. Skipping PR creation."
+          else
+            echo "Changes detected in emoji-data.js. Creating PR."
```
Comment (sy-records) [inline .github/workflows/rotate-emoji.yml:34]: Why not just commit instead of using PR?

**Review by sy-records — CHANGES_REQUESTED** [review]

File: .github/workflows/rotate-emoji.yml (hunk 5 above)
Comment (Koooooo-7) [reply .github/workflows/rotate-emoji.yml:34]: As per the previous discuss, we'd better have a review on the changes instead of commit it directly.

**Comment by Koooooo-7** [comment]
> This looks great, @Koooooo-7. A few minor naming nitpicks, but otherwise this looks good.

 I will update it.
 
> Has this been tested? If not, it seems like we could by removing a `data` property entry from `/src/core/render/emoji-data.js`, committing the change, then waiting to see if this workflow generates the expected PR.


I tested it from a fork. ( the screenshot is gotten from that).
And when we merge it into develop，we could manually trigger it for a test again.

And we don't need mock a emoji update since there already exists emoji update between our emojiData and github.

File: .github/workflows/update-emoji.yml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            git config --global user.name "GitHub Actions"
+
+            git checkout -b $branch_name
+            git add src/core/render/emoji-data.js
+            git add docs/emoji.md
+            git commit -m "update(emoji): Sync emoji data with GitHub emoji API [$current_date]"
+            unique_branch_name=$branch_name-$(git log -1 --pretty=%h)
+            git checkout -b $unique_branch_name
+            git push origin $unique_branch_name
+            gh pr create --title "$pr_title" --body "Found updated github emojis need to sync, PTAL @docsifyjs/reviewers :robot:." --base develop
```
Comment (sy-records) [inline .github/workflows/update-emoji.yml:50]: This can be simplified by using the `action`

https://github.com/stefanzweifel/git-auto-commit-action
https://github.com/thomaseizinger/create-pull-request

**Review by jhildenbiddle — APPROVED** [review]
Looks like @sy-records prefers we replace the git commands with actions, but otherwise this looks good to me.

File: .github/workflows/update-emoji.yml (hunk 6 above)
Comment (Koooooo-7) [reply .github/workflows/update-emoji.yml:50]: Thx for it.
I know those actions but according to our plain needs，using them gonna makes the action more complicated to me.
1. file change check steps and ::OUTPUT.
2. commit action on pre if-condition.
3. pr with change condition with if-condition also.

Personally, I think we could put them together without much actions dependencies for now.

**Comment by Koooooo-7** [comment]
> Looks like @sy-records prefers we replace the git commands with actions, but otherwise this looks good to me.

I will negotiate with him and do the updates id needs.

**Review by sy-records — APPROVED** [review]

**Comment by jhildenbiddle** [comment]
Sounds good, @Koooooo-7. Looks like this is ready to merge whenever you and @sy-records resolve the actions discussion. :)

Thanks for getting this up and running so quickly!
